File: prisioner/java/src/main/java/com/everis/competition/judge/Printer.java
package com.everis.competition.judge;

import static java.util.Map.Entry.comparingByValue;
import static java.util.stream.Collectors.toMap;

import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

class Printer {
	/*
	 * This class prints the score for the global results 
	 */
	private  Map<String, Integer> finalScoring = new HashMap<String, Integer>();
	private  Map<String, Integer> agesScoring = new HashMap<String, Integer>();
	private  int numberOfRounds = 0;
	private  int numberOfMatches = 0;
	
	void printSortedByAges() {
		/*
		 * This function orders the score by ages			 
		 */
		Map<String, Integer> sortByAges = agesScoring
		        .entrySet()
		        .stream()
		        .sorted(comparingByValue())
		        .collect(
		            toMap(e -> e.getKey(), e -> e.getValue(), (e1, e2) -> e2,
		                LinkedHashMap::new));

		AtomicReference<Boolean> isWinner = new AtomicReference<>();
		isWinner.set(true);
		System.out.println(Constants.MESSAGE_N_TOTAL_MATCHES+String.valueOf(numberOfMatches));
		System.out.println(Constants.MESSAGE_N_ROUNDS+String.valueOf(numberOfRounds)+" |\n");
		sortByAges.forEach((name, ages)-> {
			int points = finalScoring.get(name);
			printInnerResult(isWinner, name, points ,ages);
		});

	}

	void printSortedByPoints() {
		/*
		 * This function orders the score by points			 
		 */
		Map<String, Integer> sortByPoints = finalScoring
		        .entrySet()
		        .stream()
		        .sorted(comparingByValue(Comparator.reverseOrder()))
		        .collect(
		            toMap(e -> e.getKey(), e -> e.getValue(), (e1, e2) -> e2,
		                LinkedHashMap::new));
		
		AtomicReference<Boolean> isWinner = new AtomicReference<>();
		isWinner.set(true);
		System.out.println(Constants.MESSAGE_N_TOTAL_MATCHES+String.valueOf(numberOfMatches)+" |\n");
		System.out.println(Constants.MESSAGE_N_ROUNDS+String.valueOf(numberOfRounds)+" |\n");
		sortByPoints.forEach((name, points) -> {
			int ages = agesScoring.get(name);
			printInnerResult(isWinner, name, points ,ages);
		});
		
		
	}
	private void printInnerResult(AtomicReference<Boolean> isWinner, String name, Integer points,Integer ages) {
		int roundsParticipated = numberOfRounds;
		roundsParticipated =(roundsParticipated == 0) ? Constants.SOLO_ROUND: roundsParticipated; // Sanity check, divisor cannot be zero.
		double averageAges = (ages*1.0)/roundsParticipated;
		String outMessage = name;
		if(isWinner.get()) {
			  outMessage += Constants.MESSAGE_WINNER;
			  isWinner.set(false);
		  } 
		 if(Constants.SHOW_POINTS) {
			 outMessage = outMessage +Constants.MESSAGE_POINTS +String.valueOf(points);
		 }
		 outMessage += Constants.MESSAGE_AGES+String.valueOf(ages); 
		 // Every prisoner has participated in (n-rounds -1) 
		  outMessage +=Constants.MESSAGE_AVERAGE_AGES+String.valueOf(Math.round(averageAges));
		  System.out.println(outMessage);
	}
	
	public Printer(Map<String, Integer> finalScoring, Map<String, Integer> agesScoring, int numberOfRounds, int numberOfMatches) {
		this.finalScoring = finalScoring;
		this.agesScoring = agesScoring;
		this.numberOfRounds = numberOfRounds;
		this.numberOfMatches = numberOfMatches;
	}
	
	

}

File: prisioner/js/judge/Printer.js
var CONSTANTS = require('./Constants.js').CONSTANTS;

var Printer = function(finalScoring, agesScoring, numberOfRounds, numberOfMatches){
    this._finalScoring = finalScoring;
    this._agesScoring = agesScoring;
    this._numberOfRounds = numberOfRounds;
    this._numberOfMatches = numberOfMatches;     
    this.printSortedByAges = function(){
        var resultString = "";
        resultString +=CONSTANTS.MESSAGE_N_TOTAL_MATCHES+this._numberOfMatches+" |\n";
        resultString +=CONSTANTS.MESSAGE_N_ROUNDS+this._numberOfRounds+" |\n";
        // TODO: Refactor the format conversion after sort in another function 
        var unsortedArray = [];
        for(var key in this._agesScoring){
           unsortedArray.push({name:key, value: this._agesScoring[key]});
        }

       unsortedArray = unsortedArray.sort(function(prisoner1,prisoner2) {
           return (prisoner1.value > prisoner2.value) ? 1 : ((prisoner2.value > prisoner1.value) ? -1 : 0)
       });
       
       
       var isWinner = true;
       var outMessage="";
       for(var i in unsortedArray){
          var name = unsortedArray[i].name;
          var ages = unsortedArray[i].value;
          outMessage += this._printInnerResult(isWinner, name, ages);
          isWinner = false; // Only the first isWinner
       };
       resultString+=outMessage;
       return resultString;
    };

    this._printInnerResult = function(isWinner, name, ages){
        var outMessage = "";
        var roundsParticipated = this._numberOfRounds;
        roundsParticipated = (roundsParticipated == 0) ? CONSTANTS.SOLO_ROUND : roundsParticipated; // Sanity check
        var average_ages = Math.round((ages*1.0)/roundsParticipated);
        outMessage +=  name;
        if(isWinner==true){
         outMessage+= CONSTANTS.MESSAGE_WINNER;
         isWinner = false;
        }
        if(CONSTANTS.SHOW_POINTS)
        {
            outMessage+=CONSTANTS.MESSAGE_POINTS+this._finalScoring[name];
        }
        outMessage+= CONSTANTS.MESSAGE_AGES+ages;
        // Every prisoner has participated in (n-rounds -1) 
        outMessage += CONSTANTS.MESSAGE_AVERAGE_AGES+average_ages;
        outMessage+="\n";
        return outMessage;
    }
};



module.exports.Printer = Printer;